### PR TITLE
add least-nodes expander to cluster-autoscaler

### PR DESCRIPTION
**Pull request by com6056**
#### What this PR does / why we need it:

Adds a new `least-nodes` expander 🎉 

#### Which issue(s) this PR fixes:

No related issues

#### Does this PR introduce a user-facing change?

```release-note
Added a new `least-nodes` expander
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/expander/mostpods/mostpods.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ func (m *mostpods) BestOptions(expansionOptions []expander.Option, nodeInfo map[
 	for _, option := range expansionOptions {
 		if len(option.Pods) == maxPods {
 			maxOptions = append(maxOptions, option)
+			continue
```
Comment (com6056) [inline cluster-autoscaler/expander/mostpods/mostpods.go:40]: tiny little perf improvement

**Comment by drmorr0** [comment]
You can write/run your own expanders using the grpc expander option: https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/expander/grpcplugin

My personal opinion is that it's better for users to add new expanders using this approach, rather than adding more special-case expanders into the core autoscaler code.

File: cluster-autoscaler/expander/leastnodes/leastnodes_test.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/stretchr/testify/assert"
+
+	"k8s.io/autoscaler/cluster-autoscaler/expander"
+)
+
+func TestLeastNodes(t *testing.T) {
+	e := NewFilter()
+
+	eo0 := expander.Option{Debug: "EO0", NodeCount: 2}
```
Comment (MaciekPytel) [inline cluster-autoscaler/expander/leastnodes/leastnodes_test.go:30]: Please use a table-driven test. This test is very simple and I know a lot of older CA code uses unittests like this, but we're (slowly) pushing towards migrating to table-driven tests as those are a standard in go.

**Comment by MaciekPytel** [comment]
I generally agree with @drmorr0, but this expander seems both simple and very generic (*). As such I can imagine it being useful enough for many other users, so personally I'm ok letting it in. 

(*) Preference for large nodes is a reasonable high-level strategy, as larger nodes suffer less from "system tax" and resource fragmentation. In small clusters the downside may be lack of HA, but for a cluster with hundreds of nodes I think it is a decent strategy. And there is nothing specific to a provider or environment in the implementation, it will work anywhere as the only input it relies on is absolutely necessary for CA to be able to scale-up at all.

File: cluster-autoscaler/expander/leastnodes/leastnodes_test.go (hunk 2 above)
Comment (com6056) [reply cluster-autoscaler/expander/leastnodes/leastnodes_test.go:30]: Can do!

Comment (com6056) [reply cluster-autoscaler/expander/leastnodes/leastnodes_test.go:30]: Are there any good examples I can copy from in the repo in terms of the style to follow? I'll take a look around too

**Comment by com6056** [comment]
@MaciekPytel would you be able to run the cherry-pick command for me? Thanks! 🙏

**Comment by rljohnsn** [comment]
Would very much like to utilize this expander. Currently our similar node groups spread across AZ's are unevenly scaled. Always bumping the pool that has fewer nodes seams like the easiest logic to keep them all relatively equal.

**Comment by jacobrandall** [comment]
> @MaciekPytel would you be able to run the cherry-pick command for me? Thanks! 🙏

Also looking to use this, any chance it could get into a release soon?

**Pull request merged**